runtime/vm/MHInterpreter.cpp

```diff
@@ -1353,6 +1366,127 @@ VM_MHInterpreter::getArgSlotsBeforePosition(j9object_t argumentTypes, U_32 argPo
+
+		/* Determine the type of each argument specified by argumentIndices before copying.
+		* Note: long/double type takes 2 slots while other types take 1 slot.
+		*/
+		j9object_t argumentTypeAtIndex =
+			_objectAccessBarrier->inlineIndexableObjectReadObject(
+				_currentThread,
+				argumentTypes,
+				argumentTypeIndex);
+		J9Class *argumentClassAtIndex = J9VM_J9CLASS_FROM_HEAPCLASS(_currentThread, argumentTypeAtIndex);
+
+		if ((_vm->doubleReflectClass == argumentClassAtIndex)
+			|| (_vm->longReflectClass == argumentClassAtIndex)
+		) {
+			spCombinerSlot -= 2;
+			*(U_64*)spCombinerSlot = *(U_64*)(spFirstFilterArgSlot - argumentTypeSlots - 2);
+		} else {
+			spCombinerSlot -= 1;
+			*spCombinerSlot = *(spFirstFilterArgSlot - argumentTypeSlots - 1);
+		}
+	}
+	_currentThread->sp = spCombinerSlot; /* after loop spCombinerSlot is at the top of the combiner arguments */
+
+	return combinerHandle;
+}
+
+j9object_t
+VM_MHInterpreter::replaceReturnValueForFilterArgumentsWithCombiner()
+{
+	UDATA *bp = _currentThread->arg0EA - (sizeof(J9SFStackFrame)/sizeof(UDATA*));
+	J9SFStackFrame *frame = (J9SFStackFrame*)(bp);
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame args filterHandle PlaceHolderFrame combinerReturnValue] */
+	j9object_t filterHandle = *(j9object_t*)_currentThread->arg0EA;
+	j9object_t filterType = getMethodHandleMethodType(filterHandle);
+	j9object_t argumentTypes = getMethodTypeArguments(filterType);
+	U_32 filterArgSlots = (U_32)getMethodTypeArgSlots(filterType);
+	U_32 filterPosition = getMethodHandleFilterPosition(filterHandle);
+
+	/* Count the slot number of all the arguments before the specified filter position */
+	U_32 argumentSlots = getArgSlotsBeforePosition(argumentTypes, filterPosition);
+
+	/* Locally store the return value from the combinerHandle and determine stackslots required by the return value */
+	j9object_t combinerHandle = getCombinerHandleForFilter(filterHandle);
+	j9object_t combinerType = getMethodHandleMethodType(combinerHandle);
+	j9object_t combinerReturnType = J9VMJAVALANGINVOKEMETHODTYPE_RETURNTYPE(_currentThread, combinerType);
+	J9Class *combinerReturnTypeClass = J9VM_J9CLASS_FROM_HEAPCLASS(_currentThread, combinerReturnType);
+	UDATA combinerReturnSlots = 1;
+	UDATA combinerReturnValue0 = _currentThread->sp[0];
+	UDATA combinerReturnValue1 = 0;
+	if ((_vm->doubleReflectClass == combinerReturnTypeClass) || (_vm->longReflectClass == combinerReturnTypeClass)) {
+		combinerReturnSlots = 2;
+		combinerReturnValue1 = _currentThread->sp[1];
+	}
+
+	/* Determine the number of argument slots starting after the filterPosition */
+	U_32 remainingArgSlots = filterArgSlots - argumentSlots - combinerReturnSlots;
+
+	/* Set local variables to restore state of the _currentThread during updateVMStruct */
+	UDATA *mhPtr = UNTAGGED_A0(frame);
+
+	/* Advance past the frame and single argument (now filterHandle is at the top of the stack)) */
+	bp = (UDATA *)(((J9SFStackFrame*)(bp+1)) + 1);
```

Comment:
This assumes that the J9SFMethodTypeFrame is immediately after the PlaceHolderFrame:
```
J9SFStackFrame
MethodHandle
J9SFMethodTypeFrame
```
I don't see how this works with the extra copy of the args from the `filterArgumentsWithCombiner` method.

Reply:
I think the extra args in the stack comment is a mistake it should be `/* [... filterHandle args descriptionBytes MethodTypeFrame filterHandle PlaceHolderFrame combinerReturnValue] */` similar to what it was in the filterArgumentsWithCombinerHandle method